_posts/2022/2022-01-31-infrastructure-software-we-use-to-run-our-saas.md

```diff
@@ -0,0 +1,181 @@
+
+
+### Hosting & CI: [Heroku](https://heroku.com)
+
+Fundementally this means that we do not have the ability to break our own infrastructure. If I was to make one recommendation to anyone starting a software service it would be use a service like this where you have to configure nothing.
+
+The inflexibility of infrastructure choice means deployment remains simple and you have to work hard to insert idiosynratic complexity you don't need.
+
+You’re not totally isolated from infrastrucre concerns and you might have to architect your application in a certain way to make the most of how a Platform-as-a-Service (PaaS) Heroku works, but you are spared the overhead of managing fleets of servers or containers.
+
+They also run our PostgreSQL and Redis databases and rely on their CI & Pipelines to manage deployment.
+
+#### Scaling: [Hirefire](https://www.hirefire.io)
+
+Automatically scales our heroku dynos based on response time and queue backlog. Solid.
+
+#### Testing Parallelisation: [Knapsack Pro](https://knapsackpro.com)
+
+Added to our Heroku CI to parallelise it. The integration was very easy and it cut our builds to under 5 minutes. Most of that time is during setup: installing a browser for our system tests.
+
+### Performance, Errors & Uptime: [AppSignal](https://appsignal.com)
+
+Not as deep technically as some APM software, but at the right level for us and covers a bunch of infrastucture monitoring bases. Works nicely with Heroku. For instance the AppSignal PostgreSQL dashboard gives better insight than the (frankly poor) Heroku version.
+
+And they send out [stroopwafels](https://www.appsignal.com/waffles).
+
+
+### Application Security: [Sqreen](https://sqreen.io)
+
+One piece of software that I'm especially glad to have found. It constant reviews and protects our applications from both deliberate attacks from troublemakers and our own inadvertanty security issues.
+
+Purchased by Datadog, I guess they’ll wind that into their main product at some point.
+
+
+### Logging: [LogDNA](https://www.logdna.com)
+
+A nice UI and reasonable pricing for our volume.
+
+
+### Static Site Hosting: [Netlify](https://netlify.com)
+
+Our static sites are deployed here, we also make some use of a couple of very simple JavaScript “serverless” functions that are called from our applications.
+
+It has a similarly straightforward deployment approach to Heroku, which is one of the main reasons we use it.
+
+
+### CDN: [Fastly](https://fastly.com)
+
+Solid CDN. The main players are all much of a muchness in this arena. We're only (currently) using it for our Rails assets.
+
+
+### Payment Processing: [Stripe](https://stripe.com)
+
+Why would we use anything else to deal with our credit card transactions? Arguably without it, the business would have struggled to get off the ground.
+
+I speak as a man who previously implemented online credit card processing directly with banks. It was horrible.
+
+
+### Code Collaboration: [Github](https://github.com)
```

Comment:
*GitHub — and should be everywhere if you wanna be proper...

Reply:
```suggestion
### Code Collaboration: [GitHub](https://github.com)
```